### PR TITLE
fix(celery): stop closing prerun_span too soon to account for Celery chains scenario

### PR DESCRIPTION
**Pull request by wantsui**
We've made a few changes to handle celery context recently, including: https://github.com/DataDog/dd-trace-py/pull/10676

In particular the goal of https://github.com/DataDog/dd-trace-py/pull/10676 was to handle a scenario where a long running task may run into an exception, preventing it from closing.

Unfortunately, this scenario did not account for cases where tasks are chained and may not close until later.

See: https://github.com/DataDog/dd-trace-py/issues/11479 and https://github.com/DataDog/dd-trace-py/issues/11624

With this PR, the sample app in https://github.com/DataDog/dd-trace-py/issues/11479 would attach the celery specific span back to the root span.

I also need to add tests for the chains scenario.

Related to AIDM-494

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: tests/contrib/celery/test_chained_task.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        celery_worker_cmd.split(),
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        preexec_fn=os.setsid,
+        close_fds=True,
+        cwd=str(os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(__file__))))),
+    )
+
+    # Wait a few seconds for the celery worker to start up before trying to run tasks
+    time.sleep(5)
```
Comment (tabgok) [inline tests/contrib/celery/test_chained_task.py:25]: oof, can we use something like `celery.task.control.Inspect.stats()` instead?  Sleep-for-external-service are a great source of flakiness.

```
while celery.task.control.Inspect.stats() != worker and max_time not waited:
    sleep_a_small_amount
```

File: tests/contrib/celery/test_chained_task.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        stderr=subprocess.PIPE,
+        preexec_fn=os.setsid,
+        close_fds=True,
+        cwd=str(os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(__file__))))),
+    )
+
+    task_runner_process.wait()
+
+    # Kill the process so it starts to send traces to the Trace Agent
+    os.killpg(worker_process.pid, signal.SIGKILL)
```
Comment (tabgok) [inline tests/contrib/celery/test_chained_task.py:38]: why is this needed?

Comment (wantsui) [reply tests/contrib/celery/test_chained_task.py:38]: If the worker process doesn't get killed, the process will just run endlessly. I'm killing it so there are less logs to parse through.

File: tests/contrib/celery/test_chained_task.py (hunk 1 above)
Comment (wantsui) [reply tests/contrib/celery/test_chained_task.py:25]: Thanks! I'll look into that. Maybe that's why it's failing in CI sometimes but not locally!

File: tests/contrib/celery/test_chained_task.py (hunk 2 above)
Comment (tabgok) [reply tests/contrib/celery/test_chained_task.py:38]: ok - this won't work on windows, though (see https://stackoverflow.com/questions/28551180/how-to-kill-subprocess-python-in-windows)

File: tests/contrib/celery/test_chained_task.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+def test_task_chain_task_call_task():
+    celery_worker_cmd = "ddtrace-run celery -A tests.contrib.celery.tasks worker -c 1 -l DEBUG -n uniquename1 -P solo"
+    celery_task_runner_cmd = "ddtrace-run python tests/contrib/celery/run_tasks.py"
+    worker_process = subprocess.Popen(
+        celery_worker_cmd.split(),
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        preexec_fn=os.setsid,
+        close_fds=True,
+        cwd=str(os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(__file__))))),
```
Comment (tabgok) [inline tests/contrib/celery/test_chained_task.py:21]: do you want to split and look at [-1] ?

**Review by emmettbutler — COMMENTED** [review]
There are still some failing tests - I'll wait to review until they're passing

File: tests/contrib/celery/test_chained_task.py (hunk 2 above)
Comment (wantsui) [reply tests/contrib/celery/test_chained_task.py:38]: It looks like we call `os.killpg` in several places today: https://github.com/search?q=repo%3ADataDog%2Fdd-trace-py%20os.killpg&type=code

Is there a way to exclude some tests from running on Windows?

Comment (wantsui) [reply tests/contrib/celery/test_chained_task.py]: Update: I think I can achieve this with just the subprocess instead: https://docs.python.org/3/library/subprocess.html#subprocess.Popen.kill, which should make it more stable in windows.

I'll see how it performs in CI though.

File: tests/contrib/celery/test_chained_task.py (hunk 3 above)
Comment (wantsui) [reply tests/contrib/celery/test_chained_task.py]: I ended up just updating this so that we run the test in the current directory where the test live, instead of the convoluted way of trying to go back to the `dd-trace-py` root directory.

```
    celery_worker_cmd = "ddtrace-run celery -A tasks worker -c 1 -l DEBUG -n uniquename1 -P solo"
    celery_task_runner_cmd = "ddtrace-run python run_tasks.py"

    # The commands need to run from the directory where this test file lives
    current_directory = str(os.path.dirname(__file__))

```

Resolving this.

**Review by emmettbutler — APPROVED** [review]

**Review by tabgok — APPROVED** [review]

File: tests/contrib/celery/test_chained_task.py (hunk 2 above)
Comment (wantsui) [reply tests/contrib/celery/test_chained_task.py]: the tests pass in CI now :)

File: tests/contrib/celery/test_chained_task.py (hunk 1 above)
Comment (wantsui) [reply tests/contrib/celery/test_chained_task.py]: Resolving since this is fixed in the recent commit.
